Distributed Ensemble (MPI Support)

Robadob opened the pull request:
The implementation of MPI Ensembles within this PR is designed for each `CUDAEnsemble` to have exclusive access to all the GPUs available to it (or specified with the `devices` config). Ideally a user will launch 1 MPI worker per node, however it could be 1 worker per GPU per node.

It would be possible to use MPI shared-memory groups to identify workers on the same node and negotiate division of GPUs, and/or for some workers to become idle, however this has not been implemented.

Full notes of identified edge cases are in the below todo list.

---------------
- [x] Setup CMake
- [x] Update `CUDAEnsemble`
  - [x] Handle the 3 possible error configs
  - [x] Do something about programmatic output logs (currently each copy will provide full `RunLog`, with those not handled empty of data, checking step counter > 0 is a hack, this presumably also affects failed runs under normal error modes)
      * Each runner has their own logs in RunLog (This requires an API break, as RunLog vector does not identify run index or `RunPlan` (it was initially assumed that vector would be parallel with `RunPlanVector` used as input). *[Agreed with Paul 2023-07-26]*
      * ~~Rank 0 gets all the logs, others get empty~~
      * ~~All runners get all logs~~
  * ~~Catch and handle local MPI runs inside `CUDAEnsemble` ? (we want users to avoid using multiple MPI runners that have access to same GPU)~~
  * ~~Should we expose world_size/rank to `HostAPI`? (I don't think it's necessary)~~
- [x] Design a test case
- [x] Test on Mavericks/Waimu (single node)
- [x] Test on Bede (multi node)
- [x] Document (readme)
- [x] Document (userguide) https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/156
- [x] Update telemetry to account for MPI
- [x] Do something with ensemble example? (e.g. set `CudaEnsemble::Config().mpi=false;`)
* ~~Do we need to handle a race condition with RTC cache?~~

Closes #1073
Closes #1114

---

**Edit for visibility (by @ptheywood)** - needs to be clear in the next release notes that this includes a breaking change to the return type of `CUDAEnsemble::getLogs` from a `std::vector` to a `std::map`

Robadob commented:
I tested the render and fixed two broken link syntaxes, will merge after CI.

Robadob commented:
I've created a simple test case that will either run on a local node (if worker count <= device count) or across multiple nodes (I could probably extend this to ensure it's 1 worker per node, but that would require some MPI comms to setup the test).

The issue with MPI testing is that `MPI_Init()`, `MPI_Finalize()` can only be called once per process. With `CUDAEnsemble` auto cleaning up and triggering `MPI_Finalize()` which waits for all runners to also call it. A second MPI test case cannot be run.

Perhaps an argument for Pete's CMake test magic, as I understand that runs the test suite for each individual test. Alternative would be to add a backdoor, that tells Ensemble not to finalize when it detects tests (and add some internal finalize equivalent to ensure sync).

Requires discussion/agreement.

Robadob commented:
https://mpitutorial.com/tutorials/running-an-mpi-cluster-within-a-lan/

Setting up MPI to run across mav+waimu seems a bit involved, probably better to try Bede. I would hope the fact it works on single node is evidence that it will work though.

Robadob commented:
Had `TestMPIEnsemble.local` segfault at ~13/100 running on Waimea without mpi (at which point it should bypass MPI and just run as normal `CUDAEnsemble`. Unable to reproduce, repeating test passed. Possible rare race condition.

Robadob commented:
Happy for this to be tested on Bede and merged whilst I'm on leave. Functionality should be complete, may just want to test on Bede and refine how we wish to test it (e.g. including error handling test).

Robadob commented:
> Had `TestMPIEnsemble.local` segfault at ~13/100 running on Waimea without mpi (at which point it should bypass MPI and just run as normal `CUDAEnsemble`. Unable to reproduce, repeating test passed. Possible rare race condition.

This happened a second time. Currently just throwing it through gdb over and over to try and catch it.

Curiously this second time was directly after a recompile, so possibly only occurs when GPUs have dropped into low power state?

Robadob commented:
Caught it, race-condition when adding the run log (previous we had pre-allocated a vector so no mutex was required.

ptheywood commented:
I'll review this and test it on bede while you're on leave, and try to figure out a decent way to test it (and maybe move mpi_finalize to cleanup or similar, though again that would mean it can only be tested once).

Robadob commented:
As discussed with @ptheywood (on Slack), will move `MPI_Finalize()` to `cleanup()`, and replace with `MPI_Barrier()` (to ensure synchronisation before all workers leave the call to `CUDAEnsemble::simulate()`.

This will require adjustments to the documentation and tests.

mondus reviewed (dismissed):
Need to add MPI to requirements list in README if option is enabled.

ptheywood commented:
Also need to consider how this will behave with telemetry: flag indicating mpi use, number of ranks?, how to do the list of devices from each node etc. 

(this is a note mostly for me when I review this in the near future)

Robadob commented:
I can throw in a telemetry envelope at the final barrier if desired so rank 0 receives all gpu names.

Robadob commented:
I've now added MPI to readme requirements and ensured all tests pass with local MPI and sans MPI.

ptheywood commented:
>  Should we expose world_size/rank to HostAPI? (I don't think it's necessary)

I agree it's not neccessary to expose it ourselves, its globally available so anyone who needs it will be able to access it directly themselves (with appropriate guarding).

As there's only one rank per node (based on the docs PR) anyway, it doesn't help with uniqueness checks anyway as the rank will be the same for all simulations within a node currently anyway, so the run plan index or similar will still need using for uniqueness checks.

ptheywood commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -199,11 +211,274 @@ unsigned int CUDAEnsemble::simulate(const RunPlanVector &plans) {
-    for (unsigned int i = 0; i < TOTAL_RUNNERS; ++i) {
-        runners[i].thread.join();
-        runners[i].~SimRunner();
+    // In OMP mode, only Rank 0 increments the error counter
```
is OMP correct here?

ptheywood commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -199,11 +211,274 @@ unsigned int CUDAEnsemble::simulate(const RunPlanVector &plans) {
-        runners[i].~SimRunner();
+    // In OMP mode, only Rank 0 increments the error counter
+    unsigned int err_count = 0;
+    if (config.mpi) {
+#ifdef FLAMEGPU_ENABLE_MPI
```
This ordering means that the mpi can be enabeld in the config, but not compiled in? so there's no guarantee that somethign will happen? 

I assume its this way round so the else clause works without needing to be duplicated though?

Robadob replied:
`config.mpi` is `const false` when `FLAMEGPU_ENABLE_MPI` is not set. So the branch will be ignored by the compiler.

ptheywood commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -417,6 +764,13 @@ int CUDAEnsemble::checkArgs(int argc, const char** argv) {
+            continue;
+        }
+        // --no-mpi Disable MPI support
+        if (arg.compare("--no-mpi") == 0) {
```
I don't think `--no-mpi` is neccesary Just don't run with `mpirun -n <N>` with any value other than 1 in this case?

ptheywood commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -441,11 +795,14 @@ void CUDAEnsemble::printHelp(const char *executable) {
 #endif
-    printf(line_fmt, "", "By default, \"slow\" will be used.");
+#ifdef FLAMEGPU_ENABLE_MPI
+    printf(line_fmt, "    --no-mpi", "Disable execution using MPI");
```
--no-mpi is not neccesary imo.

ptheywood commented on src/flamegpu/simulation/detail/AbstractSimRunner.cu:
```diff
@@ -0,0 +1,114 @@
+    // Set the step config directly, to bypass validation
+    simulation->step_log_config = step_log_config;
+    simulation->exit_log_config = exit_log_config;
+    // TODO Set population?
```
Is this TODO still relevant?

Robadob replied:
Todo is taken from code originally in `CUDAEnsemble`, not required as we opt'd that ensembles must have population created via init function. Will remove/change

ptheywood commented on src/flamegpu/simulation/detail/AbstractSimRunner.cu:
```diff
@@ -0,0 +1,114 @@
+#ifdef _MSC_VER
+    std::wstringstream thread_name;
+    thread_name << L"CUDASim D" << device_id << L"T" << runner_id;
+    // HRESULT hr =
+    SetThreadDescription(this->thread.native_handle(), thread_name.str().c_str());
+    // if (FAILED(hr)) {
+    //     fprintf(stderr, "Failed to name thread 'CUDASim D%dT%u'\n", device_id, runner_id);
+    // }
+#else
+    std::stringstream thread_name;
+    thread_name << "CUDASim D" << device_id << "T" << runner_id;
+    // int hr =
+    pthread_setname_np(this->thread.native_handle(), thread_name.str().c_str());
+    // if (hr) {
+    //     fprintf(stderr, "Failed to name thread 'CUDASim D%dT%u'\n", device_id, runner_id);
+    // }
```
Lots of commented out partial code that should be cleaned up here.

Robadob replied:
Again, this is code taken from `CUDAEnsemble`. It's useful for debugging if required (in future) hence why it was originally left it. I don't think i've ever seen it fail, but we don't want to spam users regardless.

ptheywood replied:
Fair enough, just a bit of a code smell. Grand scheme its fine either way.

ptheywood commented on tests/helpers/main.cu:
```diff
@@ -22,16 +23,9 @@ GTEST_API_ int main(int argc, char **argv) {
-    }
-    // If there were more than 1 tests selected, (to exlcude bad filters and FLAMEGPU_USE_GTEST_DISCOVER related spam)
+    // Reset all cuda devices for memcheck / profiling purposes (and finalize MPI)
+    flamegpu::util::cleanup();
```
Given there are tests which call cleanup, will this work if there are MPI tests followed by cleanup tests?

Robadob replied:
I think it works so long as cleanup is not called between MPI init and the last MPI test (cant call mpi finalise before mpi init which is testable). With them all being in the same Suite, their execution should be contiguous.

ptheywood replied:
Not a problem anymore with the split test suite (assuming no one runs the main test suite with mpirun).

ptheywood commented on tests/test_cases/simulation/test_mpi_ensemble.cu:
```diff
@@ -0,0 +1,289 @@
+namespace flamegpu {
+namespace tests {
+namespace test_mpi_ensemble {
+/**
```
Do we want to split this out to a separate test binary so we can run it with more than one rank, without running every other test N times as well? 

We could configure CTest to also use mpirun -n 4 and mpirun n 1 versions separately.

ptheywood replied:
alternatively we could make it a standalone binary which just returns an appropraite error code for pass / fail, and orchestrate that via CTest if Googletest itself is causing issues.

Robadob replied:
to be discussed with Paul I guess

ptheywood replied:
Test suite has been split out.

ptheywood commented on tests/test_cases/simulation/test_mpi_ensemble.cu:
```diff
@@ -0,0 +1,289 @@
+    } else {
+        // Only rank 0 raises exception
+        EXPECT_NO_THROW(ensemble->simulate(*plans));
+        // @todo can't capture total number of successful/failed runs
```
Does this need addressing prior to merge?

Robadob replied:
Hmm, I think this should be addressable. Will have a play.

Robadob replied:
Oh this is by design, only rank 0 throws an exception/reports the number of failures etc. No point having all MPI runners pollute a log file with overlapping prints.

ptheywood commented on tests/test_cases/simulation/test_mpi_ensemble.cu:
```diff
@@ -0,0 +1,289 @@
+        model->newLayer().addHostFunction(model_step);
+    }
+    void initPlans() {
+        // 100 runs, each 10 steps which take >100ms each,
+        // therefore each sim will take >1 second.
+        // Estimate 30 seconds for full test across 4 threads
```
Do we need to do this much work in these tests? 5 tests each taking 100s when running with `mpirun -n 1` is a little tedious / is a significant increase in total test suite runtime.

Robadob replied:
Probably not, I suppose I'm just playing safe for poor latency. Probably another argument for splitting MPI tests into distinct suite.

ptheywood replied:
This would still be nice to reduce the total test time, 5 tests each taking upto 100s seems excessive. 

Could either reduce the sleep time per test, or reduce the total numeber of tests to be say 10 per MPI rank so it is consistent in time regardless of mpi world size?.

Robadob replied:
I think I prefer the latter idea. Will do

ptheywood commented on README.md:
```diff
@@ -175,6 +176,7 @@ cmake --build . --target all
 | `FLAMEGPU_RTC_DISK_CACHE`            | `ON`/`OFF`                  | Enable/Disable caching of RTC functions to disk. Default `ON`.                                             |
 | `FLAMEGPU_VERBOSE_PTXAS`             | `ON`/`OFF`                  | Enable verbose PTXAS output during compilation. Default `OFF`.                                             |
 | `FLAMEGPU_CURAND_ENGINE`             | `XORWOW` / `PHILOX` / `MRG` | Select the CUDA random engine. Default `XORWOW`                                                            |
+| `FLAMEGPU_ENABLE_MPI`                | `ON`/`OFF`                  | Enable MPI support for distributed CUDAEnsembles. Default `OFF`.                                           |
```
The restriction on <= one mpi rank per GPU per node (?) needs to be very clearly spelled out in the readme (and the docs).

Robadob replied:
It's recommended 1 MPI rank per node, FGPU can handle dividing up multiple GPUs (and this reduces comms).

Don't recall how multi gpu/rank is handled, should be noted in docs but will address here too.

Robadob replied:
Have said MPI worker needs exclusive access to gpus, each 1 per node.

ptheywood reviewed (changes requested):
A few comments. 

Additionally, the `ensemble` example needs some modification, as running it with MPI enabled means it's internal validation doesn't work. 

```
mpirun -n 4 ./bin/Release/ensemble 
CUDAEnsemble completed 100 runs successfully!
Ensemble init: 120, calculated init 450
Ensemble result: 10484775152, calculated result 40244135200
Ensemble init: 116, calculated init 450
Ensemble result: 11866340560, calculated result 40244135200
Ensemble init: 106, calculated init 450
Ensemble result: 9640392448, calculated result 40244135200
Ensemble init: 108, calculated init 450
Ensemble result: 8252627040, calculated result 40244135200
```

I've only ran the tests on a single GPU machine so far, so only a single MPI rank. 

I'll test on a multigpu single node and multi-node at a later point.

Robadob commented:
> Do you know which MPI / GCC / CUDA you compiled with previously where this worked?

Would be whatever my bashrc on waimu default to I guess

Robadob commented:
> as running it with MPI enabled means it's internal validation doesn't work.

That's why there's a disable mpi config option ;)

Robadob commented on include/flamegpu/simulation/detail/AbstractSimRunner.h:
```diff
@@ -0,0 +1,193 @@
+                                                             reinterpret_cast<char*>(&t.device_id) - t_ptr,
+                                                             reinterpret_cast<char*>(&t.runner_id) - t_ptr,
+                                                             reinterpret_cast<char*>(&t.exception_string) - t_ptr };
+            constexpr MPI_Datatype array_of_types[count] = {MPI_UNSIGNED, MPI_UNSIGNED, MPI_UNSIGNED, MPI_CHAR};
```
My waimu setup has been
gcc: 10.5.0
cuda 12.2
mpi 3.3.2 (it says HYDRA on `mpirun --version`, but eh)

Robadob replied:
Your changes builds fine.

Robadob commented on include/flamegpu/simulation/CUDAEnsemble.h:
```diff
@@ -22,6 +23,20 @@ struct RunLog;
  */
 class CUDAEnsemble {
  public:
+#ifdef FLAMEGPU_ENABLE_MPI
```
I don't really understand what you mean.

The ifdef isn't strictly required, there's little harm (other than confusion) for the below `EnvelopeTag` enum being available everywhere.

ptheywood replied:
The comment was meant to be on the enum line I expect.

`enum class` enums are type safe - i.e. you can't set it to `0`/`1` / some other enum that happens to map to `0`/`1`, you have to provide it an `EnvelopeTag`, or cast the value. 

They are just nicer IMO. 

I think I was probably asking if there was a reason e.g. do they get fed into MPI which expects the underlying int, which would require the casting from the enum class to int or not (given that MPI is a C library not c++ (since the c++ interface was removed in/after 2.0? as it caused more problems than it sovled)

Robadob replied:
Don't think im aware of `enum class`.

ptheywood replied:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#enum3-prefer-class-enums-over-plain-enums

Robadob commented on README.md:
```diff
@@ -78,6 +78,7 @@ Optionally:
   + With `setuptools`, `wheel`, `build` and optionally `venv` python packages installed
 + [swig](http://www.swig.org/) `>= 4.0.2` for python integration
   + Swig `4.x` will be automatically downloaded by CMake if not provided (if possible).
++ MPI (e.g. [MPICH](https://www.mpich.org/), [OpenMPI](https://www.open-mpi.org/)) for distributed ensemble support
```
Waimu is on 3.x, so just need to test 2.x somewhere/how.

Robadob replied:
Use MPI functions

```
MPI_Type_create_struct()
MPI_Type_commit()
MPI_Initialized()
MPI_Init_thread()
MPI_Comm_rank()
MPI_Comm_size()
MPI_Iprobe()
MPI_Recv()
MPI_Send()
MPI_Barrier()
MPI_Probe()
MPI_Get_count()
MPI_Finalized()
MPI_Finalize()
```

MPI_Type_commit() first appears in 3.0 spec best I can tell.

Quick ctrl+f suggest all the rest exist in 3.0 spec too (albeit this isn't to say prototypes match).

Robadob replied:
But open-mpi docs seem to say its in mpi 2.0.4 so who knows.

https://www.open-mpi.org/doc/v2.0/man3/MPI_Type_commit.3.php

ptheywood replied:
I only managed to test with 3.0+, with even some 3.0 builds failing due to GCC being too new.
We can probably just leave it unspecified and if anyone tries to use MPI 2.x in 2023 they probably have other issues.

Robadob commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -216,32 +491,99 @@ unsigned int CUDAEnsemble::simulate(const RunPlanVector &plans) {
+        if (config.telemetry) {
+            // All ranks should notify rank 0 of their GPU devices
+            if (world_rank == 0) {
+                int bufflen = 2048;
```
Nope, but at a guess will be long enough for any exception messages. The code below is supposed to resize it if the buffer is too short, but reviewing it it's bad it doubles the length without checking it's suitable.

Robadob replied:
Will change to 256, appears to be size of string for device name in docs

Robadob commented on tests/test_cases/simulation/test_mpi_ensemble.cu:
```diff
@@ -0,0 +1,289 @@
+            gpuErrchk(cudaGetDeviceCount(&deviceCount));
+            if (deviceCount < world_size) {
+                MPI_Finalize();
+                GTEST_SKIP() << "Skipping single-node MPI test, world size (" << world_size << ") exceeds GPU count (" << deviceCount << "), this would cause test to stall.";
```
I think the stall was a hypothesis tbh, as init/sync things wouldn't match up.

Robadob replied:
Yes it does stall.

The code directly above this basically for purposes of a single node test, split's the GPUs between the MPI workers. But if you have more workers than GPUs. One of more workers will be left without a GPU and hence can't start, so the first MPI sync hit will never be reached by those workers causing the stall.

As we recommend users only use 1 MPI worker per node (or simply every worker has unique access to it's GPUs), this seems acceptable. It's merely a good way to allow us to perform local testing without a complex setup to divide GPUs.

ptheywood replied:
Still concerned about the stall potentially being triggered by users in real code, but the tests skipping in this case is OK, although it would be nice to be able to test the mpi calls in a single GPU machine, even if that is a likely-to-fail scenario for sufficiently large models/ensembles.

Robadob replied:
Whilst what you suggest would be technically possible to make possible via the test suite. We're already constrained to running these tests on Linux, which we have two multi-gpu boxes regularly sitting idle. Hence, I'm not currently bothered enough to address it. (We already saw that MPI on a single node is not a good equivalent to multi-node [e.g. the problems we had on Bede], so testing should really be performed on any system that MPI will be used).

ptheywood replied:
There's no fundamental differentce between mpi comms within and between machines, just the latency will be higher (and not using MPI might not be the most performant implementation). 

The bede mpi issues are unrelated to the MPI calls themselves, and more with the mpi version/install config (afiak). I need to dig back through bede related comms to see if they were similar issues encountered by others (I know there were MPI issues that another RSE was investigating, but need to look into it).

`mpirun -n 2` in a single GPU machine can be its own issue to go on the backlog, but I still want to get my head around the potential deadlock prior to approval.

Robadob replied:
> mpirun -n 2 in a single GPU machine can be its own issue to go on the backlog, but I still want to get my head around the potential deadlock prior to approval.

If we do want a python test, unelss there's an mpi package the world size/rank can't be probed so any test would need to support that.

ptheywood replied:
Opened #1146. 

This scenario will need thoroguhly documenting though if it could be triggered by users.

Robadob replied:
The stall in the test suite only occurs because it refuses to allocate runners to an ensemble if they exceed gpu count. Hence the syncs within the ensemble would never be hit.

A user not messing about with mpi ranks etc, will not hit a stall.

Robadob commented:
To be discussed:
* Get rid of `--no-mpi`? (Pete doesn't like it)
* Split out MPI tests into a separate suite?
* Add an example 1 worker per node SLURM script to docs

Robadob commented:
Agreed

* Get rid of --no-mpi
* Split out MPI tests into a separate suite
* MPI file specific CI to test multiple MPI versions
* add slurm script to docs.
* ensemble example fix

Robadob commented:
> MPI file specific CI to test multiple MPI versions

I've created a new MPI workflow. However, as expected, it fails to install specific versions of mpich/openmpi via apt-get. Will wait for @ptheywood's return to advise on best method to make them available (my natural next step would be build from source).

> add slurm script to docs.

I have semi-successfully ran the mpi test suite on Bede.

```sh
#!/bin/bash

# Generic options:

#SBATCH --account=bdshe03 # Run job under project <project>
#SBATCH --time=0:10:0         # Run for a max of 10 mins

# Node resources:

#SBATCH --partition=gpu    # Choose either "gpu" or "infer" node type
#SBATCH --nodes=2          # Resources from a two nodes
#SBATCH --gres=gpu:1       # 1 GPUs per node

# Run commands:

# 1ppn == 1 process per node
bede-mpirun --bede-par 1ppn ./build/bin/Release/tests_mpi
```
 Produces the intermingled log
 
 ```
Running main() from /users/robadob/fgpu2/tests/helpers/main.cu
[0;32m[==========] [mRunning 5 tests from 1 test suite.
[0;32m[----------] [mGlobal test environment set-up.
[0;32m[----------] [m5 tests from TestMPIEnsemble
[0;32m[ RUN      ] [mTestMPIEnsemble.success
Running main() from /users/robadob/fgpu2/tests/helpers/main.cu
[0;32m[==========] [mRunning 5 tests from 1 test suite.
[0;32m[----------] [mGlobal test environment set-up.
[0;32m[----------] [m5 tests from TestMPIEnsemble
[0;32m[ RUN      ] [mTestMPIEnsemble.success
[gpu030.bede.dur.ac.uk:696907] pml_ucx.c:291  Error: Failed to create UCP worker
[gpu031.bede.dur.ac.uk:2817740] pml_ucx.c:291  Error: Failed to create UCP worker
[0;32m[       OK ] [mTestMPIEnsemble.success (56576 ms)
[0;32m[ RUN      ] [mTestMPIEnsemble.success_verbose
[0;32m[       OK ] [mTestMPIEnsemble.success (56500 ms)
[0;32m[ RUN      ] [mTestMPIEnsemble.success_verbose
[0;32m[       OK ] [mTestMPIEnsemble.success_verbose (50446 ms)
[0;32m[ RUN      ] [mTestMPIEnsemble.error_off
[0;32m[       OK ] [mTestMPIEnsemble.success_verbose (50446 ms)
[0;32m[ RUN      ] [mTestMPIEnsemble.error_off
[0;32m[       OK ] [mTestMPIEnsemble.error_off (50467 ms)
[0;32m[ RUN      ] [mTestMPIEnsemble.error_slow
[0;32m[       OK ] [mTestMPIEnsemble.error_off (50467 ms)
[0;32m[ RUN      ] [mTestMPIEnsemble.error_slow
[0;32m[       OK ] [mTestMPIEnsemble.error_slow (50463 ms)
[0;32m[ RUN      ] [mTestMPIEnsemble.error_fast
[0;32m[       OK ] [mTestMPIEnsemble.error_slow (50462 ms)
[0;32m[ RUN      ] [mTestMPIEnsemble.error_fast
[0;32m[       OK ] [mTestMPIEnsemble.error_fast (6057 ms)
[0;32m[----------] [m5 tests from TestMPIEnsemble (214011 ms total)

[0;32m[----------] [mGlobal test environment tear-down
[0;32m[==========] [m5 tests from 1 test suite ran. (214011 ms total)
[0;32m[  PASSED  ] [m5 tests.
[0;32m[       OK ] [mTestMPIEnsemble.error_fast (6057 ms)
[0;32m[----------] [m5 tests from TestMPIEnsemble (213934 ms total)

[0;32m[----------] [mGlobal test environment tear-down
[0;32m[==========] [m5 tests from 1 test suite ran. (213934 ms total)
[0;32m[  PASSED  ] [m5 tests. 
 ```
Of note:

* The tests passed on both MPI workers
* "Error: Failed to create UCP worker", not clear what this mean or the impact (googling suggests might be related to infiniband), or how to resolve.
* Bede has it's own bespoke `bede-mpirun` command which takes custom args along with mpirun args. Is this standard? It kind of harms our plan to provide a sample SLURM script.

Robadob commented:
Output from updated ensemble example using `mpirun -n 2` on waimu (I didn't do any special hacks to make sure GPUs are unique but it still worked).

```
rob@waimea:~/FLAMEGPU2/build/bin/Debug$ mpirun -n 2 ./ensemble
CUDAEnsemble completed 100 runs successfully!
Ensemble init: 218, calculated init 218
Ensemble result: 22162315712, calculated result 22162315712
Local MPI runner completed 51/100 runs.
Ensemble init: 232, calculated init 232
Ensemble result: 18081819488, calculated result 18081819488
Local MPI runner completed 49/100 runs.
```

Robadob commented:
## Status
* Need some input from @ptheywood as to how to setup MPI CI (in particular, how to install specific versions)
* Need to discuss the Bede failures:
  * Do we understand "Error: Failed to create UCP worker" does it matter?
  * Do the additional failures when using `mpirun` rather than `bede-mpirun` make sense?
  * How does this affect plans to document with a slurm script.

mondus commented:
Suggest to try with `mvapich2` from [Bede docs](https://bede-documentation.readthedocs.io/en/latest/software/libraries/mpi.html).

Robadob commented:
Document how to run test suite with mpirun

Robadob commented:
> Suggest to try with `mvapich2` from [Bede docs](https://bede-documentation.readthedocs.io/en/latest/software/libraries/mpi.html).

When build with `mvapich2`, and executed using `bede-mpirun` this error is received (~3 attempts with slight changes).

```
*** An error occurred in MPI_Init_thread
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
```

Google suggests it's an MPI misconfigure problem.

The `mvapich2` `mpirun` docs aren't great, so it doesn't have a convenient parameter like `openmpi` for 1 process per node. Can't workout required commands to bypass bede-mpirun.

ptheywood commented:
I've added a warning in CMake if using MPI and CMake < 3.20.1 to address #1114, which outputs the following.

```
CMake Warning at src/CMakeLists.txt:569 (message):
  CMake < 3.20.1 may result in link errors with FLAMEGPU_ENABLE_MPI=ON for
  some MPI installations.  Consider using CMake >= 3.20.1 to avoid linker
  errors.
```

ptheywood commented:
Stanage `znver3` (i.e. for the GPU nodes) include `OpenMPI`, so we can use Stanage's A100 and H100 nodes for an x86_64 single-node upto 4 GPU MPI bench as it currently stands.

From an A100/H100 node, i.e. the following successfully conifgured and compiled from a H100 node:

```bash
module load OpenMPI/4.1.4-GCC-11.3.0 GCC/11.3.0 CUDA/11.8.0  CMake/3.24.3-GCCcore-11.3.0
mkdir -p build-11-8-mpi
cd build-11-8-mpi
cmake .. -DCMAKe_CUDA_ARCHITECTURES="80;90" -DFLAMEGPU_BUILD_TESTS=ON -DFLAMEGPU_ENABLE_MPI=ON
cmake --build . --target tests_mpi -j `nproc`
```

I then ran the MPI test suite using a single rank, with a single GPU in my interactive session

```bash
mpirun -n 1 bin/Release/tests_mpi
```

This ran successfully, but took quite a while. Might be worth toning theses tests down so they don't take as long when only using a single rank?

```console
Running main() from /users/ABCDEFG/code/flamegpu/FLAMEGPU2/tests/helpers/main.cu
[==========] Running 5 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 5 tests from TestMPIEnsemble
[ RUN      ] TestMPIEnsemble.success
[       OK ] TestMPIEnsemble.success (101671 ms)
[ RUN      ] TestMPIEnsemble.success_verbose
[       OK ] TestMPIEnsemble.success_verbose (101131 ms)
[ RUN      ] TestMPIEnsemble.error_off
[       OK ] TestMPIEnsemble.error_off (100341 ms)
[ RUN      ] TestMPIEnsemble.error_slow
[       OK ] TestMPIEnsemble.error_slow (100338 ms)
[ RUN      ] TestMPIEnsemble.error_fast
[       OK ] TestMPIEnsemble.error_fast (10333 ms)
[----------] 5 tests from TestMPIEnsemble (413816 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test suite ran. (413817 ms total)
[  PASSED  ] 5 tests.

```

Trying to run with 4 processes on a single GPU and 20 CPU cores of an H100 node requires `--oversubscribe` due to the current stanage configuration (only one mpi slot available), but that configuration skips a number of tests due to the stall, and then mpirun reports the error as the google test proccesses which skip return a non zero error code

```console
/users/ABCDEFG/code/flamegpu/FLAMEGPU2/tests/test_cases/simulation/test_mpi_ensemble.cu:109: Skipped
Skipping single-node MPI test, world size (4) exceeds GPU count (1), this would cause test to stall.
/users/ABCDEFG/code/flamegpu/FLAMEGPU2/tests/test_cases/simulation/test_mpi_ensemble.cu:109: Skipped
Skipping single-node MPI test, world size (4) exceeds GPU count (1), this would cause test to stall.
/users/ABCDEFG/code/flamegpu/FLAMEGPU2/tests/test_cases/simulation/test_mpi_ensemble.cu:109: Skipped
Skipping single-node MPI test, world size (4) exceeds GPU count (1), this would cause test to stall.
/users/ABCDEFG/code/flamegpu/FLAMEGPU2/tests/test_cases/simulation/test_mpi_ensemble.cu:109: Skipped
Skipping single-node MPI test, world size (4) exceeds GPU count (1), this would cause test to stall.
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[33200,1],0]
  Exit code:    1
--------------------------------------------------------------------------
```

Robadob commented:
> This ran successfully, but took quite a while. Might be worth toning theses tests down so they don't take as long when only using a single rank?

I think bede runs with 2 gpus were taking ~3 minutes, 1 gpu should be taking ~6 minutes.

Should be trivial to add a constant to scale the time I guess.

ptheywood commented:
re: CI most MPI installs don't provide binary packages, and most linux distro's only package a single version of each MPI implementation.

I'm prototyping github action step(s) which install MPI from apt if specified or from source elsewhere (to avoid spamming long-running CI by pushing to this branch). Once it's sorted I'll add it to this branch.

ptheywood commented:
MPI CI is pasing for all openmpi's, and MPICH's from source. 

MPICH from apt is failing at link time. This appears to be related to `-flto` which is enabled for host object compilation (and passed to the host compiler for cuda objects), but it is not being passed at link time? 

This is the only build which is adding `flto`, so presumably its an implicit option coming from the mpich installation somehow. I may be able to repro this locally?

ptheywood commented:
Installing `libmpich-dev` on my ubuntu 22.04 install reproduces the error. 

The MPICH distributed via ubuntu / debian is the source of the lto flags, as shown by the following 

```console
$ mpicxx -compile-info
g++ -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -I/usr/include/x86_64-linux-gnu/mpich -L/usr/lib/x86_64-linux-gnu -lmpichcxx -lmpich
```

Not yet sure on how to resolve this in a way that will allow this build for end users. 

Within my `CMakeCache.txt`, these flags are in `MPI_CXX_COMPILE_OPTIONS`, an advanced internal cache variable. 

Potentially I could configure with this explicitly set to nothing, i.e. `-DMPI_CXX_COMPILE_OPTIONS=""` which may work, but that might break other mpi installations so isn't something we'd want to programatically set. It might also still not result in usable binaries anyway if this is actaully required.

```console 
MPI_CXX_COMPILE_OPTIONS:STRING=-flto=auto;-ffat-lto-objects;-flto=auto
``` 

I could probably detect this at configure time and warn/error about it in CMake?
Otherwise enabling LTO via CMake the "propper cmake" (`INTERPROCEDURAL_OPTIMIZATION`) way might work (I'm skeptical), but might also enable device lto.

ptheywood commented:
It is possible to detect and warn about -flto at configuration time, e.g. in src/CMakeLists.txt after the 3.20.1 warning:

```
        # If the MPI installation brings in -flto (i.e. Ubuntu 22.04 libmpich-dev), warn about it and suggest a reconfiguration.
        if(MPI_CXX_COMPILE_OPTIONS MATCHES ".*\-flto.*")
            message(WARNING
                " MPI_CXX_COMPILE_OPTIONS contains '-flto' which is likely to result in linker errors. \n"
                " Consider an alternate MPI implementation which does not embed -flto,\n"
                " Or reconfiguring CMake with -DMPI_CXX_COMPILE_OPTIONS=\"\" if linker error occur.")
        endif()
```

```console
$ cmake .. 
-- -----Configuring Project: flamegpu-----
-- CUDA Architectures: 86
-- RapidJSON found. Headers: /home/ptheywood/code/flamegpu/FLAMEGPU2/build-mpi/_deps/rapidjson-src/include
-- flamegpu version 2.0.0-rc.1+eed4987d
CMake Warning at src/CMakeLists.txt:576 (message):
   MPI_CXX_COMPILE_OPTIONS contains '-flto' which is likely to result in linker errors. 
   Consider an alternate MPI implementation which does not embed -flto,
   Or reconfiguring CMake with -DMPI_CXX_COMPILE_OPTIONS="" if linker error occur.

```

ptheywood commented on .github/workflows/MPI.yml:
```diff
@@ -0,0 +1,258 @@
+  run:
+    # Default to using bash regardless of OS unless otherwise specified.
+    shell: bash
+
+# @todo - add a job/step which validates the version is of the correct format (for pre releases) and that the git tag matches the CMake/header version(s) 
+
+# Several jobs with some dependencies between them.
+# + Thorough Ubuntu builds
+#   + MPI 2.x+
```
```suggestion
```

@todo comment is copy pasted from a scripot, not relevant. 

Also not a thorough ubuntu build, so probably just better to delete it all.

ptheywood commented on .github/workflows/MPI.yml:
```diff
@@ -0,0 +1,258 @@
+        # CUDA_ARCH values are reduced compared to wheels due to CI memory issues while compiling the test suite.
+        cudacxx:
+          - cuda: "12.0"
+            cuda_arch: "50-real;90-real;90-virtual;"
```
For this CI we can just use single arch, not min max like in the thorough build to speed it up. 
```suggestion
            cuda_arch: "50-real;"
```

ptheywood commented on include/flamegpu/simulation/detail/SimLogger.h:
```diff
@@ -34,7 +34,7 @@ class SimLogger {
      * @param _export_exit_time If true exit log time will be exported
      */
-    SimLogger(const std::vector<RunLog> &run_logs,
+    SimLogger(const std::map<unsigned int, RunLog> &run_logs,
```
This might be enough of an API break it will need mentioning in the changelog?

Robadob replied:
This is internal (`detail`).

ptheywood commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -463,7 +812,7 @@ void CUDAEnsemble::setExitLog(const LoggingConfig &exitConfig) {
     exit_log_config = std::make_shared<LoggingConfig>(exitConfig);
 }
-const std::vector<RunLog> &CUDAEnsemble::getLogs() {
+const std::map<unsigned int, RunLog> &CUDAEnsemble::getLogs() {
```
This will need to be in the changelog too. Probably worth diffing the include dir when this PR is good to merge to get the API changes.

ptheywood commented:
I've been back through the changes since I last reviewed this, leaving comments as neccesary. Will aim to do some more testing of this on a multi-node system (Bede) tomorrow if possible, and get my head around the potential stall situations just to make sure I understand it and it's not a wider problem. 

Just a few more things that need small tweaks that hadn't been addressed yet otherwise (making the tests faster etc).

ptheywood commented:
From the Todo list for this pr:

>  Should we expose world_size/rank to HostAPI? (I don't think it's necessary)

No, this is accessible by MPI for anyone that really wants / needs it

---

We should probably check this works from python too though prior to merge, and maybe add a test case for that.

Robadob commented:
> Just a few more things that need small tweaks that hadn't been addressed yet otherwise (making the tests faster etc).

I think I've addressed all your points that came through my emails.

> We should probably check this works from python too though prior to merge, and maybe add a test case for that.

I can probably get to that Friday.

ptheywood commented:
# tl;dr

testing on Stanage using an interactive 2 a100 session:

+ Running without mpirun/mpiexec fails due to mpi detecting srun but not being built for it. Ideally this should just run fine without, but not sure how feasible that is on all systems.
+ Ensemble example run count is reported as 99/100 when using 1 rank
+ Using 2/4 ranks runs with `--oversubscribe` and the count is correct. Unclear which ranks use which GPUs however.
+ The test suite does not behave correctly with `mpirun -n 2`. It runs all tests twice independently, rather than running each test once but with 2 ranks. This might be a google test limitation, so 1 non google test test per binary and a ctest run might be the solution.
+ The test suite does not behave correctly with `mpiurun -n 1` with several test failures

Commit hash `d529d6e81d7e6e4a7dfab56e4b235052dfc3700e`

---

On a Stanage, using 1/2 of an a100 node (2 GPUs, 24 cores) interactively :

```bash
# Get an interactive session
srun --partition=gpu --qos=gpu --gres=gpu:a100:2 --mem=164G --cpus-per-task 24 --pty bash -i
```

Then in the interactive job (i.e. on an AMD cpu core, not the login intel CPU)

```bash
# Load Dependencies
module load CUDA/11.8.0 OpenMPI/4.1.4-GCC-11.3.0 GCC/11.3.0 CMake/3.24.3-GCCcore-11.3.0

# Build dir
mkdir -p build-mpi-cu118-gcc-113-ompi414
cd build-mpi-cu118-gcc-113-ompi414
# Configure
cmake .. -DCMAKE_CUDA_ARCHITECTURES="80;90" -DFLAMEGPU_ENABLE_MPI=ON -DFLAMEGPU_ENABLE_NVTX=ON -DFLAMEGPU_BUILD_TESTS=ON
# Compile
cmake --build . -j `nproc`
```

## Ensemble Example

Running the ensemble example without explicitly using `mpirun` fails with an MPI error. This could be a mpi configuration thing, but ideally the binary should be usable without MPI, and just not use MPI if not requested (I'm not 100% sure this is possible for all MPI applications, given mpirun in non standard, so detecting if mpi is requested might not be possible all the time)

```console
$ ./bin/Release/ensemble
[gpu04.pri.stanage.alces.network:33719] OPAL ERROR: Unreachable in file pmix3x_client.c at line 112
--------------------------------------------------------------------------
The application appears to have been direct launched using "srun",
but OMPI was not built with SLURM's PMI support and therefore cannot
execute. There are several options for building PMI support under
SLURM, depending upon the SLURM version you are using:

  version 16.05 or later: you can use SLURM's PMIx support. This
  requires that you configure and build SLURM --with-pmix.

  Versions earlier than 16.05: you must use either SLURM's PMI-1 or
  PMI-2 support. SLURM builds PMI-1 by default, or you can manually
  install PMI-2. You must then build Open MPI using --with-pmi pointing
  to the SLURM PMI library location.

Please configure as appropriate and try again.
--------------------------------------------------------------------------
*** An error occurred in MPI_Init_thread
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
[gpu04.pri.stanage.alces.network:33719] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
```

Using mpirun (or mpiexec) with a single rank runs successfully, but the completed message is missing a run (or more, seen 95 as well)

```console
$ mpirun -n 1 ./bin/Release/ensemble
CUDAEnsemble completed 100 runs successfully!
Ensemble init: 441, calculated init 441
Ensemble result: 40144235200, calculated result 40144235200
Local MPI runner completed 99/100 runs.
```

Using 2 ranks (2 GPUs) errors with the stanage configuration from an interactive job, as mpi only beleives one slot is available. This might just be how I requsted the job. However using the suggested options it can be ran (see below)

```console
$ mpirun -n 2 ./bin/Release/ensemble
--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 2
slots that were requested by the application:

  ./bin/Release/ensemble

Either request fewer slots for your application, or make more slots
available for use.

A "slot" is the Open MPI term for an allocatable unit where we can
launch a process.  The number of slots available are defined by the
environment in which Open MPI processes are run:

  1. Hostfile, via "slots=N" clauses (N defaults to number of
     processor cores if not provided)
  2. The --host command line parameter, via a ":N" suffix on the
     hostname (N defaults to 1 if not provided)
  3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
  4. If none of a hostfile, the --host command line parameter, or an
     RM is present, Open MPI defaults to the number of processor cores

In all the above cases, if you want Open MPI to default to the number
of hardware threads instead of the number of processor cores, use the
--use-hwthread-cpus option.

Alternatively, you can use the --oversubscribe option to ignore the
number of available slots when deciding the number of processes to
launch.
--------------------------------------------------------------------------
```

Using `--oversubscribe`` works, and the MPI runner counts add up.

```console
$ mpirun --oversubscribe -n 2 ./bin/Release/ensemble
CUDAEnsemble completed 100 runs successfully!
Ensemble init: 213, calculated init 213
Ensemble result: 17859919488, calculated result 17859919488
Local MPI runner completed 45/100 runs.
Ensemble init: 237, calculated init 237
Ensemble result: 22384215712, calculated result 22384215712
Local MPI runner completed 55/100 runs.
```

Using 4 threads and `--oversubscribe` also works, even though there are only 2 GPUs.

```console
$ time mpiexec --oversubscribe -n 4 ./bin/Release/ensemble 
CUDAEnsemble completed 100 runs successfully!
Ensemble init: 126, calculated init 128
Ensemble result: 11977340560, calculated result 11977540560
Local MPI runner completed 29/100 runs.
Ensemble init: 70, calculated init 69
Ensemble result: 6409861632, calculated result 6409761632
Local MPI runner completed 16/100 runs.
Ensemble init: 141, calculated init 140
Ensemble result: 11424857856, calculated result 11424757856
Local MPI runner completed 28/100 runs.
Ensemble init: 113, calculated init 113
Ensemble result: 10432075152, calculated result 10432075152
Local MPI runner completed 27/100 runs.

real    0m4.540s
user    0m2.145s
sys     0m16.240s
```

## Test suite

The test suite does not behave as intended when running with MPIrun, it just runs all the tests twice with one rank each, rather than running each test once, but using multiple processes.
This might be a google test limitation, in which case we might need to make each mpi test it's own binary, and orchestrate them via ctest (we can use categories to make it easy to just run the mpi tests)

```console
$ mpirun --oversubscribe -n 2 ./bin/Release/tests_mpi 
Running main() from /users/ac1phey/code/flamegpu/FLAMEGPU2/tests/helpers/main.cu
[==========] Running 5 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 5 tests from TestMPIEnsemble
[ RUN      ] TestMPIEnsemble.success
Running main() from /users/ac1phey/code/flamegpu/FLAMEGPU2/tests/helpers/main.cu
[==========] Running 5 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 5 tests from TestMPIEnsemble
[ RUN      ] TestMPIEnsemble.success
[       OK ] TestMPIEnsemble.success (10767 ms)
[ RUN      ] TestMPIEnsemble.success_verbose
[       OK ] TestMPIEnsemble.success (10772 ms)
[ RUN      ] TestMPIEnsemble.success_verbose
[       OK ] TestMPIEnsemble.success_verbose (10029 ms)
[ RUN      ] TestMPIEnsemble.error_off
[       OK ] TestMPIEnsemble.success_verbose (10029 ms)
[ RUN      ] TestMPIEnsemble.error_off
[       OK ] TestMPIEnsemble.error_off (10030 ms)
[ RUN      ] TestMPIEnsemble.error_slow
[       OK ] TestMPIEnsemble.error_off (10031 ms)
[ RUN      ] TestMPIEnsemble.error_slow
[       OK ] TestMPIEnsemble.error_slow (10030 ms)
[ RUN      ] TestMPIEnsemble.error_fast
[       OK ] TestMPIEnsemble.error_slow (10030 ms)
[ RUN      ] TestMPIEnsemble.error_fast
[       OK ] TestMPIEnsemble.error_fast (6020 ms)
[----------] 5 tests from TestMPIEnsemble (46879 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test suite ran. (46879 ms total)
[  PASSED  ] 5 tests.
[       OK ] TestMPIEnsemble.error_fast (6020 ms)
[----------] 5 tests from TestMPIEnsemble (46883 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test suite ran. (46883 ms total)
[  PASSED  ] 5 tests.
```

Using a single rank test fail:

```console
$ mpirun -n 1  ./bin/Release/tests_mpi 
Running main() from /users/ac1phey/code/flamegpu/FLAMEGPU2/tests/helpers/main.cu
[==========] Running 5 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 5 tests from TestMPIEnsemble
[ RUN      ] TestMPIEnsemble.success
[       OK ] TestMPIEnsemble.success (5951 ms)
[ RUN      ] TestMPIEnsemble.success_verbose
[       OK ] TestMPIEnsemble.success_verbose (5105 ms)
[ RUN      ] TestMPIEnsemble.error_off
/users/ac1phey/code/flamegpu/FLAMEGPU2/tests/test_cases/simulation/test_mpi_ensemble.cu:201: Failure
Expected equality of these values:
  err_count
    Which is: 0
  1u
    Which is: 1
/users/ac1phey/code/flamegpu/FLAMEGPU2/tests/test_cases/simulation/test_mpi_ensemble.cu:204: Failure
Value of: errors.find("Warning: Run 10 failed on rank ") != std::string::npos
  Actual: false
Expected: true
[  FAILED  ] TestMPIEnsemble.error_off (5106 ms)
[ RUN      ] TestMPIEnsemble.error_slow
/users/ac1phey/code/flamegpu/FLAMEGPU2/tests/test_cases/simulation/test_mpi_ensemble.cu:232: Failure
Value of: errors.find("Warning: Run 10 failed on rank ") != std::string::npos
  Actual: false
Expected: true
[  FAILED  ] TestMPIEnsemble.error_slow (5107 ms)
[ RUN      ] TestMPIEnsemble.error_fast
[  FAILED  ] TestMPIEnsemble.error_fast (5106 ms)
[----------] 5 tests from TestMPIEnsemble (26377 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test suite ran. (26377 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] TestMPIEnsemble.error_off
[  FAILED  ] TestMPIEnsemble.error_slow
[  FAILED  ] TestMPIEnsemble.error_fast

 3 FAILED TESTS
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[56666,1],0]
  Exit code:    1
--------------------------------------------------------------------------

```

ptheywood commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -199,11 +211,274 @@ unsigned int CUDAEnsemble::simulate(const RunPlanVector &plans) {
+                // Print progress to console
+                if (config.verbosity >= Verbosity::Default && next_run_reported != next_run) {
+                    const int progress = static_cast<int>(next_run) - static_cast<int>(TOTAL_RUNNERS * world_size);
+                    fprintf(stdout, "\rCUDAEnsemble progress: %d/%u", progress < 0 ? 0 : progress, static_cast<unsigned int>(plans.size()));
```
I dislike the usage of carriage returns it breaks if a model includes any stdout output. 

They make logging the output to file store messy info, and if a printf is included in the model at all it breaks output (admitidily with concurrent ensembles this will be out of order anyway)

Not going to fight its inclusion, just logging my dislike.

Robadob replied:
Find me the relevant code to detect the stream type to allow output file to be mutated and i'll make the change. Iirc last time I tried, it didn't work.

ptheywood replied:
Actually given the use-case for MPI (distribtued batch jobs) then just getting rid of the carraige returns makse more sense than having them at all. 

I.e. it's always going to be batch jobs on hpc systems, where logging to disk will happen, so no carraige returns is the better option. 

---

should be `isatty` from `unistd.h` iirc. it's part of the posix spec though not c/c++, so would need guarding appropriately, but depending on how programs are executed it might not report a differnt value than expected (I think).

Robadob replied:
fair point.

Robadob commented:
Re: incorrect progress. I think this is a consequence of the reporting logic

```
const int progress = static_cast<int>(next_run) - static_cast<int>(TOTAL_RUNNERS * world_size);
```

It would never report higher than `TOTAL_RUNS-TOTAL_RUNNERS` with a single MPI rank, as it's reporting the id of the newly assigned job - the number of runners. But will jump to exit early as soon as `next_run` exceeds `TOTAL_RUNS`.

Regardless, I've changed it so the `printf` instead reports which job index has been assigned to which rank (and got rid of the `\r`).

Robadob commented:
I've reworking progress printing, fixed bugs that were causing temperamental test failures, added a few extra tests for better coverage and added a single pyflamegpu capable MPI test (e.g. one that is very limited, but won't break if ran via MPI).

ptheywood commented on include/flamegpu/simulation/detail/MPIEnsemble.h:
```diff
@@ -0,0 +1,108 @@
+    /**
+     * @return Retrieve the local world rank from MPI
+     */
+    int getWorldRank();
```
Any reason these methods are private but underlying variables are public scoped? 

Would make more sense for the mehtods to be pub and the variables to be private.

Robadob replied:
The vars are const, so them being public is fine.

The methods actually trigger MPI to initialise MPI if required before returning the value. They aren't trivial getters, they are initialisers.

ptheywood commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -441,11 +630,14 @@ void CUDAEnsemble::printHelp(const char *executable) {
     printf(line_fmt, "    --standby", "Allow the machine to enter standby during execution");
 #endif
-    printf(line_fmt, "", "By default, \"slow\" will be used.");
+#ifdef FLAMEGPU_ENABLE_MPI
+    printf(line_fmt, "    --no-mpi", "Disable execution using MPI");
+#endif
```
--no-mpi has been removed(?), but still in help output. 

```suggestion
```

ptheywood commented:
Code generally looks good now, still need to re-run things in a distributed setting etc. 

One final idea for change, is to use `MPI_Comm_split_type` within the cuda ensemble to auto-assign gpus for multiple ranks in a single multigpu node, much like how this is handled in the existing tests to support single-rank testing.

I.e. 

+ Split the communicator to get the number of mpi ranks on per node
+ Get the number of gpus for the node the current rank is executing on.
+ If the user specified the gpus for the ensemble, do what they said unless it will break things
+ If not, load balance the gpus within the node across the mpi ranks
  + If group_size == 1, use all gpus
  + If group_size > 1 && group_size < num_gpus, load balance (do integer division to get num per gpu, assignt that until run out of gpus. so balancedunless say 3 ranks and 4 gpus, in which case it'd be a 1:1:2 split)
  + If group_size > num_gpus, use 1 gpu per mpi rank, and the rest do nothing (maybe emit a warning). Or it could be an error

This is mainly so that systems where it is not possible to request 1 mpi rank per node (or not easily) can be supported. Using 1 rank per node would still (likely) be the best performance. 

(This would also make for an interesting benchmark / would let some benchmarking scenarios be covered). 

I might do this myself (leaving the comment so I don't forget)

ptheywood commented:
I've prototyped creating a new MPI communicator which will only invovle mpi ranks which have been assigned a GPU, to prevent errors if users request more mpi ranks per node than are available, with those ranks doing nothing. 

E.g. 

```c++
int max_participating_ranks_per_node = 3; // pretend each rank has 3 gpus / 3 mpi ranks participating
    int color = local_rank < max_participating_ranks_per_node ? 0 : 1;

    if (MPI_Comm_split(MPI_COMM_WORLD, color, world_rank, &comm_participating) != MPI_SUCCESS) {
        fprintf(stderr, "Error creating communicator\n");
        MPI_Abort(MPI_COMM_WORLD, 1);
    }
    int participating_size, participating_rank;
    MPI_Comm_size(comm_participating, &participating_size);
    MPI_Comm_rank(comm_participating, &participating_rank);
```


Launching with 4 ranks each on 2 nodes, pretending there are 3 gpus each then shows the correct comms. 


```
main rank 0: recieving (6 - 1) messages?
messages from: 2, 1, 4, 5, 6, 
```

ptheywood commented:
MPI tests in debug builds are currently failing on mavericks due to captured output not matching what the test contains. This was the case prior to my device selection changes but hadn't been noticed (probably just hadn't been ran in debug mode for a while).

More ranks than GPUs ensemble example runs are working, but test suite is not behaving (I've likely broken assumptions in the tests now I've removed the check on the number of mpi ranks / device setting within the test suite)

Also not yet implemented what to do when devices are specified.

ptheywood commented:
Release test suite all sorted now. 

@Robadob - if you could skim the commits I've added to make sure you're happy enough with it that'd be appreciated. I've not touched the user-facing interface at all (other than changing implicit behaviour, that will need a tweak to the docs api).

Still need to:

+ [ ] Fix running the test suite in debug (this was an issue before my changes)
+ [ ] Python... 
  + Should make sure it works in python fine still / in general, i.e. add atleast one explicit test 
  + Probably need to tweak the python `conftest.py` so it doesn't telemetry from each rank, but that will be grim.
  + No idea if pytest will be happy when executed using mpi or not. 

Otherwise should be there now (I think I've got all my debug etc removed).

ptheywood commented:
Test suite is fixed, python test suite is crashing in test_logging.py + maybe others for an mpi build. Had missed 2 printf's too. 

doing a non-mpi build to figure out if its mpi build specific or not before attempting further debugging (I have a feeling there's a breaking change in here to do with a method return type which is probably the culprit). 

Given the absence of a python_mpi tests sub suite, this might be a bit of effort. Might need to change conftest stuff for telemetry too.

ptheywood commented:
5 ptest failures in the non-MPI build fails the python test suite with which need resolving.

2 have been fixed elsewhere, so just needs a rebase. Log related ones are an api break? 

```
FAILED ../tests/python/codegen/test_codegen_integration.py::GPUTest::test_gpu_codegen_function_condition - AttributeError: '_SpecialForm' object has no attribute '__name__'
FAILED ../tests/python/codegen/test_codegen_integration.py::GPUTest::test_gpu_codegen_simulation - AttributeError: '_SpecialForm' object has no attribute '__name__'
FAILED ../tests/python/io/test_logging.py::LoggingTest::test_CUDAEnsembleSimulate - AttributeError: 'int' object has no attribute 'getStepLog'
FAILED ../tests/python/simulation/test_cuda_ensemble.py::TestCUDAEnsemble::test_setExitLog - AttributeError: 'int' object has no attribute 'getExitLog'
FAILED ../tests/python/simulation/test_cuda_ensemble.py::TestCUDAEnsemble::test_setStepLog - AttributeError: 'int' object has no attribute 'getStepLog'
```

Robadob commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -150,14 +151,26 @@ unsigned int CUDAEnsemble::simulate(const RunPlanVector& plans) {
         }
     }
+#ifdef FLAMEGPU_ENABLE_MPI
+    // if MPI is enabled at compile time, and mpi is enabled in the config, use the MPIEnsemble method to assign devices balanced across ranks
```
```suggestion
    // if MPI is enabled at compile time, use the MPIEnsemble method to assign devices balanced across ranks
```

config mpi is compile time const

ptheywood replied:
Based on this. 

https://github.com/FLAMEGPU/FLAMEGPU2/blob/cdad83a2dd9c34e2ebefc3a7fbea740180e7a62e/src/flamegpu/simulation/CUDAEnsemble.cu#L66

Robadob commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -150,14 +151,26 @@ unsigned int CUDAEnsemble::simulate(const RunPlanVector& plans) {
     }
+#ifdef FLAMEGPU_ENABLE_MPI
+    // if MPI is enabled at compile time, and mpi is enabled in the config, use the MPIEnsemble method to assign devices balanced across ranks
+    if (mpi != nullptr) {
```
This condition should always be true within the IFDEF

ptheywood replied:
Based on this. 

https://github.com/FLAMEGPU/FLAMEGPU2/blob/cdad83a2dd9c34e2ebefc3a7fbea740180e7a62e/src/flamegpu/simulation/CUDAEnsemble.cu#L66

Robadob commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -172,6 +185,36 @@ unsigned int CUDAEnsemble::simulate(const RunPlanVector& plans) {
 
+    // If there are no devices left (and mpi is not being used), we need to error as the work cannot be executed.
+#ifdef FLAMEGPU_ENABLE_MPI
+    if (devices.size() == 0 && mpi == nullptr) {
```
`mpi==nullptr` will always evaluate `false`?

ptheywood replied:
Based on this. 

https://github.com/FLAMEGPU/FLAMEGPU2/blob/cdad83a2dd9c34e2ebefc3a7fbea740180e7a62e/src/flamegpu/simulation/CUDAEnsemble.cu#L66

Robadob commented on src/flamegpu/simulation/CUDAEnsemble.cu:
```diff
@@ -172,6 +185,36 @@ unsigned int CUDAEnsemble::simulate(const RunPlanVector& plans) {
+
+#ifdef FLAMEGPU_ENABLE_MPI
+    // Once the number of devices per rank is known, we can create the actual communicator to be used during MPI, so we can warn/error as needed.
+    if (mpi != nullptr) {
```
Redundant condition?

ptheywood replied:
Based on this. 

https://github.com/FLAMEGPU/FLAMEGPU2/blob/cdad83a2dd9c34e2ebefc3a7fbea740180e7a62e/src/flamegpu/simulation/CUDAEnsemble.cu#L66

ptheywood commented:
Tests now fixed by skipping, exposing `pyflamegpu.MPI` indicating if MPI was enabled for the build or not. 

Ran on bede using 
+ openmpo with 2 nodes, 1 gpu per node and one rank per node, ensemble and tests were fine, on T4 and V100 nodes
+ openmpi wwith 2 nodes, 2 gpus per node and 2 ranks per node on T4's, this errored due to an mpi error which has an open issue (power + infinband + gpus), so not an "us" problem.
+ mvapich, 2 infer nodes, 2 gpus per node, 2 ranks per node. ran fine.

ptheywood commented:
Will remove the ternary and associated checks for it to be nullptr'd tomorrow, then should be goood to merge I think. (+ update the docs pr to reflect the change to the requirement of 1 rank per node, instead to be up to 1 rank per gpu (or a warning will be given)

ptheywood commented:
Tweaked the final suggestions, so think this is good to go now. 

Not sure Rob can review this PR though, so I will have to approve (or Paul)?
